apps/checkout/src/providers/ErrorsProvider.tsx

```diff
@@ -0,0 +1,57 @@
+import { useGetFormErrorsFromApiErrors } from "@/lib/utils";
+import { PropsWithChildren, useState } from "react";
+import { FieldErrors } from "react-hook-form";
+import createSafeContext from "./createSafeContext";
+
+export type ApiErrors<TFormData> = Array<{
+  field: keyof TFormData;
+  code: "REQUIRED" | "INVALID";
+  message: string;
+}>;
+
+export type Errors<TFormData> = FieldErrors<TFormData>;
+
+export type ErrorsContextConsumerProps<TFormData = any> = {
+  errors: Errors<TFormData>;
+  hasErrors: boolean;
+  setErrorsFromApi: (apiErrors: ApiErrors<TFormData>) => void;
+  setErrors: (errors: FieldErrors<TFormData>) => void;
+  clearErrors: () => void;
+};
+
+interface ErrorsProviderProps<TFormData> {
+  apiErrors: ApiErrors<TFormData>;
+}
+
+export const [useErrorsContext, Provider] =
+  createSafeContext<ErrorsContextConsumerProps>();
+
+export const ErrorsProvider = function <TFormData>({
```

Comment:
What's context used for in this case? It looks like this might as well be a hook, what's shared state used for?

Reply:
For once to avoid passing props a couple of levels down, currently userAddresses wraps its children in the provider and passes down api errors that are needed by it's childrens children. Its child UserAddressSection also has 2 different api errors that it adds to that pole. Could use a hook for each of those, but since only one form can be displayed in this section at one time, it's much easier to assume that we only need one type of errors at the time (so if displaying "address create" section we care for address create errors etc.). Seems like less boilerplate (cause we don't have to keep addressCreateErrors + userAddressUpdateErrors + user address create errors + shippingAddress update errors), less logic (cause we clear/replace errors in one place and not 3) and less props passing. fair enough?